Kernel/Process.cpp

```diff
@@ -687,7 +688,41 @@ int Process::do_exec(String path, Vector<String> arguments, Vector<String> envir
         // Okay, here comes the sleight of hand, pay close attention..
         auto old_regions = move(m_regions);
         m_regions.append(move(executable_region));
+
+        AK::ArmedScopeGuard rollback_regions_guard([&](){
+            m_page_directory = move(old_page_directory);
+            ASSERT(&current->process() == this);
+            MM.enter_process_paging_scope(*this);
+            executable_region = m_regions.take_first();
+            m_regions = move(old_regions);
+        });
+
         loader = make<ELFLoader>(region->vaddr().as_ptr(), metadata.size);
+        auto interpreter_path = loader->interpreter();
+
+        if (interpreter_path.has_value()) {
+            dbgprintf("%s(%d) do_exec(%s): Using program interpreter %s\n", m_name.characters(), m_pid, path.characters(), interpreter_path.value().characters());
```

Comment:
Should this share code with `find_shebang_interpreter_for_executable`? Maybe that doesn't make sense.

Reply:
Not sure.. I'm not super thrilled about how it presently does all the "whoops, wrong elf binary" processing with SMAP disabled too. If there's a better place to read the elf header and check for PT_INTERP without mapping the entire file, it wasn't immediately obvious to me.

Reply:
Not really. The SMAP disabler is supposed to be temporary until we can go through all the code and figure out ways to avoid it. If we make the ELF image region supervisor-only, we should be able to access it even with SMAP protection enabled. I think that's the way forward there.

The ELF code is still way too trusting btw. We need to start checking every dang offset it gives us before using it to index into anything, and failing gracefully if something looks like shenanigans.

Reply:
Okay SMAP protections are now enabled throughout `exec()`.

This has got me thinking though.. ELF loading for `exec()` purposes can be made so much simpler. If we assume that the program headers follow the file header (and call shenanigans if they don't), we could simply read the first page (4KB) of the ELF and walk the program headers without using `ELFLoader` or `ELFImage` at all.

This would drastically simplify in-kernel program loading. What do you think?

Reply:
Well, it sounds great until you remember that the section header table is stored at the end of the binary instead of the beginning. So, you'd lose symbolication the way it currently happens because that's where the .symtab and .strtab are. 

So that change would probably want to come along with either userspace stack unwinding. I think glibc sets up unwind tables as part of libc init? Once the dynamic loader is actually in it'll break symbolication too, since all the symbols for the program won't be in the elf image the kernel initially loaded (thanks DT_NEEDED). So it's something that would need addressed anyway.

Reply:
Symbolication seems like something we should be doing in userspace either way.

Anyways, this is unrelated to this particular PR so I'll stop yapping about it here :)

Reply:
I think the latest push does essentially what you suggested (or at least starts to). Symbolication is still another issue though.